In [2]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import balanced_accuracy_score, confusion_matrix, classification_report

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [3]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
df = pd.read_csv("Resources/lending_data.csv")

# Review the DataFrame
df.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0

In [4]:
#Finding out how many rows are duplicated data
df.duplicated()


0        False
1        False
2        False
3        False
4        False
         ...  
77531    False
77532     True
77533     True
77534    False
77535     True
Length: 77536, dtype: bool

In [5]:
#dropping the duplicated rows
df = df.drop_duplicates()

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [6]:
# Separate the data into targets and features
# Separate the y variable, the labels
y = df["loan_status"]

# Separate the X variable, the features
X = df.drop('loan_status', axis=1)

In [7]:
# Review the y variable Series
y.head()

0    0
1    0
2    0
3    0
4    0
Name: loan_status, dtype: int64

In [8]:
# Review the X variable DataFrame
X.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100  
3                 1       22700  
4                 1       23000

### Step 3: Check the balance of the labels variable (`y`) by using the `value_counts` function.

In [12]:
# Check the balance of our target values
y.value_counts()

0    3653
1    1576
Name: loan_status, dtype: int64

### Step 4: Split the data into training and testing datasets by using `train_test_split`.

In [14]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)

---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [16]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
clf = LogisticRegression(random_state=1).fit(X_train, y_train)


### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [18]:
# Make a prediction using the testing data
y_pred = clf.predict(X_test)

In [20]:
y_test.value_counts()

0    909
1    399
Name: loan_status, dtype: int64

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [19]:
# Print the balanced_accuracy score of the model
balanced_accuracy_score(y_test, y_pred)

0.8937359901403674

In [22]:
# Generate a confusion matrix for the model
confusion_matrix(y_test, y_pred)

array([[832,  77],
       [ 51, 348]])

In [21]:
# Print the classification report for the model
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.94      0.92      0.93       909
           1       0.82      0.87      0.84       399

    accuracy                           0.90      1308
   macro avg       0.88      0.89      0.89      1308
weighted avg       0.90      0.90      0.90      1308



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** Although this model has decent accuracy with successfully predicting healthy loans, the accuracy was subpar for predicting high-risk loans; arguably a more important risk factor when a business is considering handing out loans. The driver of this low accuracy with high-risk loan predictions was due to the data set not being well balanced by a 2-to-1 margin.  There are 2 processes that will increase accuracy of high-risk predictions: a) going back to the original data set and removing duplicated data points (there are over 70k duplicated data points that will be removed from the data in a second-pass over the beginning lines of code in this notebook), and b) re-sampling the data with a RandomOverSampler in the cells below.

---

## Predict a Logistic Regression Model with Resampled Training Data

### Step 1: Use the `RandomOverSampler` module from the imbalanced-learn library to resample the data. Be sure to confirm that the labels have an equal number of data points. 

In [29]:
# Import the RandomOverSampler module form imbalanced-learn
from imblearn.over_sampling import RandomOverSampler

# Instantiate the random oversampler model
# # Assign a random_state parameter of 1 to the model
ros = RandomOverSampler(random_state=1)

# Fit the original training data to the random_oversampler model
X_res, y_res = ros.fit_resample(X_train, y_train)
X_res.head(20)

loan_size  interest_rate  borrower_income  debt_to_income  \
0      6600.0          5.931            36400        0.175824   
1     18800.0         11.104            85100        0.647474   
2      9300.0          7.088            47300        0.365751   
3      8600.0          6.801            44600        0.327354   
4     18400.0         10.930            83500        0.640719   
5      7100.0          6.157            38500        0.220779   
6     16400.0         10.115            75800        0.604222   
7      8200.0          6.623            42900        0.300699   
8     20800.0         11.964            93200        0.678112   
9     19200.0         11.300            86900        0.654776   
10    13000.0          8.630            61800        0.514563   
11     9600.0          7.214            48500        0.381443   
12    19900.0         11.584            89600        0.665179   
13    18000.0         10.784            82100        0.634592   
14    21300.0         12.183            95200        0.684874   
15    16100.0          9.950            74200        0.595687   
16    10400.0          7.526            51400        0.416342   
17    21300.0         12.183            95300        0.685205   
18     9600.0          7.185            48200        0.377593   
19    18100.0         10.812            82400        0.635922   

    num_of_accounts  derogatory_marks  total_debt  
0                 1                 0        6400  
1                12                 2       55100  
2                 3                 0       17300  
3                 3                 0       14600  
4                12                 2       53500  
5                 1                 0        8500  
6                10                 2       45800  
7                 2                 0       12900  
8                14                 3       63200  
9                12                 2       56900  
10                7                 1       31800  
11                4                 0       18500  
12               13                 2       59600  
13               11                 2       52100  
14               14                 3       65200  
15                9                 2       44200  
16                4                 1       21400  
17               14                 3       65300  
18                4                 0       18200  
19               11                 2       52400

In [23]:
# Count the distinct values of the resampled labels data
y_res.value_counts()

0    2744
1    2744
Name: loan_status, dtype: int64

### Step 2: Use the `LogisticRegression` classifier and the resampled data to fit the model and make predictions.

In [24]:
# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
clf_res = LogisticRegression(random_state=1).fit(X_res, y_res)

# Make a prediction using the testing data
y_pred = clf_res.predict(X_test)

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [25]:
# Print the balanced_accuracy score of the model 
balanced_accuracy_score(y_test, y_pred)

0.9409580055749935

In [26]:
# Generate a confusion matrix for the model
confusion_matrix(y_test, y_pred)

array([[829,  80],
       [ 12, 387]])

In [27]:
# Print the classification report for the model
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.99      0.91      0.95       909
           1       0.83      0.97      0.89       399

    accuracy                           0.93      1308
   macro avg       0.91      0.94      0.92      1308
weighted avg       0.94      0.93      0.93      1308



### Step 4: Answer the following question

**Question:** How well does the logistic regression model, fit with oversampled data, predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** Using the oversampled data the logistic regression model correctly predicts 829 healthy and 389 high-risk loans.  This shifts the balance in the possibilty of falsely classifying a low-risk loan as a high risk loan considerably while maintaining 94% accuracy. 